sycl/include/CL/sycl/detail/image_accessor_util.hpp

```diff
@@ -144,67 +257,431 @@ void writePixel(vec<T, 4> Pixel, T *Ptr, image_channel_order Order) {
 
 // Converts read pixel data into return datatype based on the channel type of
 // the image.
-// TODO: Change this method to use the conversion rules as given in the OpenCL
+// Conversion rules used as given in the OpenCL
 // Spec section 8.3. The conversion rules may be handled differently for each
 // return datatype - float, int32, uint32, half. ImageChannelType is passed to
 // the function to use appropriate conversion rules.
-template <typename ChannelType, typename RetDataType,
-          typename = detail::enable_if_t<
-              (detail::is_contained<
-                  RetDataType, type_list<cl_int, cl_float, cl_uint>>::value)>>
+
+template <typename ChannelType>
 void convertReadData(vec<ChannelType, 4> PixelData,
                      image_channel_type ImageChannelType,
-                     vec<RetDataType, 4> &RetData) {
-  RetData.x() = (RetDataType)PixelData.x();
-  RetData.y() = (RetDataType)PixelData.y();
-  RetData.z() = (RetDataType)PixelData.z();
-  RetData.w() = (RetDataType)PixelData.w();
+                     vec<uint, 4> &RetData) {
+
+  switch (ImageChannelType) {
+  case image_channel_type::unsigned_int8:
+  case image_channel_type::unsigned_int16:
+  case image_channel_type::unsigned_int32:
+    RetData = PixelData.template convert<uint>();
+    break;
+  default:
+    // OpenCL Spec section 6.12.14.2 does not allow reading uint4 data from an
+    // image with channel datatype other than unsigned_int8,unsigned_int16 and
+    // unsigned_int32.
+    assert(!"Datatype of read data - cl_uint4 is incompatible with the "
```

Comment:
Shouldn't it better to make it exception?

Reply:
yes.